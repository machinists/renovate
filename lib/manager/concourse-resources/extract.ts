import is from '@sindresorhus/is';
import { load } from 'js-yaml';
import { logger } from '../../logger';
import { id as dockerVersioning } from '../../versioning/docker';
import { getDep } from '../dockerfile/extract';
import type { PackageDependency, PackageFile } from '../types';
import type { ConcourseDependency } from './types';

function getConcourseDep({
  registry,
  repository,
  tag,
}: {
  registry: string;
  repository: string;
  tag: string;
}): PackageDependency {
  const dep = getDep(`${registry}${repository}:${tag}`, false);
  dep.replaceString = tag;
  dep.versioning = dockerVersioning;
  dep.autoReplaceStringTemplate =
    '{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}';
  return dep;
}

/**
 * Recursively find all supported dependencies in the yaml object.
 *
 * @param parsedContent
 */
function findDependencies(
  parsedContent: Record<string, unknown> | ConcourseDependency,
  packageDependencies: Array<PackageDependency>
): Array<PackageDependency> {
  if (!parsedContent || typeof parsedContent !== 'object') {
    return packageDependencies;
  }

  if (is.string(parsedContent.type) && parsedContent.type === 'docker-image') {
    const currentItem = parsedContent.source;

    let registryMirror: string = currentItem.registry_mirror;
    registryMirror = registryMirror ? `${registryMirror}/` : '';
    const repository = String(currentItem.repository);
    const tag = String(currentItem.tag);
    packageDependencies.push(
      getConcourseDep({ repository, tag, registryMirror })
    );
  }

  return packageDependencies;
}

export function extractPackageFile(content: string): PackageFile {
  let parsedContent: Record<string, unknown> | ConcourseDependency;
  try {
    // a parser that allows extracting line numbers would be preferable, with
    // the current approach we need to match anything we find again during the update
    // TODO: fix me (#9610)
    parsedContent = load(content, { json: true }) as any;
  } catch (err) {
    logger.debug({ err }, 'Failed to parse helm-values YAML');
    return null;
  }
  try {
    const deps = findDependencies(parsedContent, []);
    if (deps.length) {
      return { deps };
    }
  } catch (err) /* istanbul ignore next */ {
    logger.error({ err }, 'Error parsing helm-values parsed content');
  }
  return null;
}
