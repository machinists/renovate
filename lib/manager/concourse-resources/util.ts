import { hasKey } from '../../util/object';
import { regEx } from '../../util/regex';
import { ConcourseDockerImageDependency } from './types';

const parentKeyRe = regEx(/image$/i);

/**
 * Type guard to determine whether a given partial Concourse values.yaml object potentially
 * defines a Concourse Docker dependency.
 *
 * There is no exact standard of how Docker dependencies are defined in Concourse
 * values.yaml files (as of February 26th 2021), this function defines a
 * heuristic based on the most commonly used format in the Concourse charts:
 *
 * image:
 *   repository: 'something'
 *   tag: v1.0.0
 * renovateImage:
 *   repository: 'something'
 *   tag: v1.0.0
 */
export function matchesConcoursePipelineDockerHeuristic(
  parentKey: string,
  data: unknown
): data is ConcourseDockerImageDependency {
  return (
    parentKeyRe.test(parentKey) &&
    data &&
    typeof data === 'object' &&
    hasKey('repository', data) &&
    hasKey('tag', data)
  );
}

export function matchesConcoursePipelineInlineImage(
  parentKey: string,
  data: unknown
): data is string {
  return parentKeyRe.test(parentKey) && data && typeof data === 'string';
}
