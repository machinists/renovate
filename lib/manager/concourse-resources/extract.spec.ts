import { loadFixture } from '../../../test/util';
import { extractPackageFile } from './extract';

const concourseDefaultPipelineInitValues = loadFixture(
  'default_pipeline_init_pipeline.yaml'
);

describe('manager/concourse-pipeline/extract', () => {
  describe('extractPackageFile()', () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });
    it('returns null for invalid yaml file content', () => {
      const result = extractPackageFile('nothing here: [');
      expect(result).toBeNull();
    });
    it('returns null for empty yaml file content', () => {
      const result = extractPackageFile('');
      expect(result).toBeNull();
    });
    it('returns null for no file content', () => {
      const result = extractPackageFile(null);
      expect(result).toBeNull();
    });
    it('extracts from pipeline.yaml correctly', () => {
      const result = extractPackageFile(concourseDefaultPipelineInitValues);
      expect(result).toMatchSnapshot({
        deps: [
          {
            currentValue: 'v0.19.1',
            depName: 'teliaoss/github-pr-resource',
          },
        ],
      });
    });
  });
});
