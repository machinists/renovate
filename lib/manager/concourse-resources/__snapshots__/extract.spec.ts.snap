// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager/concourse-resources/extract extractPackageFile() extracts from pipeline.yaml correctly 1`] = `
Object {
  "deps": Array [
    Object {
      "autoReplaceStringTemplate": "{{newValue}}{{#if newDigest}}@{{newDigest}}{{/if}}",
      "currentDigest": undefined,
      "currentValue": "v0.19.1",
      "datasource": "docker",
      "depName": "teliaoss/github-pr-resource",
      "replaceString": "v0.19.1",
      "versioning": "docker",
    },
  ],
}
`;
